control/controller.go

```diff
@@ -580,37 +516,59 @@ func (q *Query) setErr(err error) {
 
 	// Close the ready channel to report that no results
 	// will be sent.
-	close(q.ready)
+	close(q.results)
+}
+
+// pump will read from the executing query results and pump the
+// results to our destination.
+// When there are no more results, then this will close our own
+// results channel.
+func (q *Query) pump(exec flux.Query) {
+	defer close(q.results)
+
+	done := q.currentCtx.Done()
+	for {
+		select {
+		case res, ok := <-exec.Results():
+			if !ok {
+				return
+			}
+			q.results <- res
+		case <-done:
+			// Signal to the underlying executor that the query
+			// has been canceled. Usually, the signal on the context
+			// is likely enough, but this explicitly signals just in case.
+			exec.Cancel()
+
+			// Set the done channel to nil so we don't do this again
+			// and we continue to drain the results.
+			done = nil
```

Comment:
If the query has been canceled, shouldn't we break out of the loop?

Reply:
The reason for this is because cancel signals that the loop should be broken out of and since we are wrapping the query from the program that is being executed, we need to rely on when that query says we are done and not ourselves. So this signals to the inner query that the results have been canceled, but continues pumping the results.